### PR TITLE
fix: regression in loading LDAP users policy mappings

### PR DESCRIPTION
**Pull request by harshavardhana**
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license] (https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
fix: regression in loading LDAP users policy mappings

## Motivation and Context
LDAP users are stored as STS users, we need to 
load their policy mappings appropriately.

Fixes a regression caused by #17994

## How to test this PR?
```sh
#!/bin/bash

pkill -9 minio
rm -rf ~/test*

export MC_HOST_svc=http://minio123:minio123@localhost:9000
export MC_HOST_site=http://minioadmin:minioadmin@localhost:9000

./minio server --console-address ":9002" ~/test{1...4} &

sleep 1

mc idp ldap add site server_addr=localhost:1389 server_insecure=on \
	lookup_bind_dn=cn=admin,dc=min,dc=io lookup_bind_password=admin \
	user_dn_search_base_dn=dc=min,dc=io user_dn_search_filter="(uid=%s)" \
	group_search_base_dn=ou=swengg,dc=min,dc=io group_search_filter="(&(objectclass=groupOfNames)(member=%d))"

mc admin service restart site/

sleep 1

mc idp ldap policy attach site consoleAdmin --user='uid=dillon,ou=people,ou=swengg,dc=min,dc=io'
mc admin user svcacct add site dillon --access-key=minio123 --secret-key=minio123

mc mb site/testbucket

mc admin user svcacct ls site "uid=dillon,ou=people,ou=swengg,dc=min,dc=io"
mc admin group list site

mc idp ldap policy entities site

mc ls svc/

mc admin service restart site/

sleep 1

mc ls svc/
```

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] Fixes a regression #17994
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

File: cmd/iam-store.go [hunk 1]
```diff
@@ -1620,7 +1619,11 @@ func (store *IAMStoreSys) UserNotificationHandler(ctx context.Context, accessKey
 	}
 
 	if userType != svcUser {
-		err = store.loadMappedPolicy(ctx, accessKey, userType, false, cache.iamUserPolicyMap)
+		if userType == stsUser {
+			err = store.loadMappedPolicy(ctx, accessKey, userType, false, cache.iamSTSPolicyMap)
+		} else {
+			err = store.loadMappedPolicy(ctx, accessKey, userType, false, cache.iamUserPolicyMap)
+		}
```
Comment (donatello) [inline cmd/iam-store.go:1626]: 1. This is the main fix - a notification to update a user should load the policy for sts policy into the right map. IIUC the bug is happening in distributed setup where the policy mapping is not loaded on a node other than the node where the policy update was made.

**Comment by cesnietor** [comment]
can we add tests for this?

**Review by poornas — APPROVED** [review]

File: cmd/iam-store.go [hunk 2]
```diff
@@ -965,7 +958,13 @@ func (store *IAMStoreSys) PolicyDBUpdate(ctx context.Context, name string, isGro
...
 	if !isGroup {
 		if userType == stsUser {
-			mp = cache.iamSTSPolicyMap[name]
+			var ok bool
+			mp, ok = cache.iamSTSPolicyMap[name]
+			if !ok {
+				// Attempt to load parent user mapping for STS accounts
+				store.loadMappedPolicy(context.TODO(), name, stsUser, false, cache.iamSTSPolicyMap)
+				mp = cache.iamSTSPolicyMap[name]
+			}
```
Comment (donatello) [inline cmd/iam-store.go:967]: 2. There may still be an issue here. The bug being fixed here is that the stsPolicyMap does not contain the STS user's policy. This is fixed in the notification handler to load the mapping properly. However, there is a problem when the notification is missed by a node and the policy mapping is being updated (i.e. already exists and is being changed). In that case, a stale mapping would be present, and the policy DB update would not be correct.

Comment (harshavardhana) [reply cmd/iam-store.go:967]: Isn't that expected to be fixed via refresh() routine? @donatello

**Review by donatello — APPROVED** [review]

**Pull request merged**